### PR TITLE
[#291] Add function for CRC32C checksum

### PR DESCRIPTION
**Pull request by shikharish**

### CODE REVIEW

File: src/include/security.h [hunk 1]
```diff
@@ -139,6 +139,31 @@ int pgmoneta_generate_string_hmac_sha256_hash(char* key, int key_length, char* v
...
+int pgmoneta_create_crc32c_buffer(char* buffer, size_t size, uint32_t* crc_buf);
+
+/**
+ * Generate CRC32C for a buffer using AVX instructions
+ * @param buffer The buffer.
+ * @param size The size of the buffer.
+ * @param crc_buff The hash value.
+ * @return 0 upon success, otherwise 1.
+ */
+int pgmoneta_create_crc32c_buffer_avx(char* buffer, size_t size, uint32_t* crc_buf);
```
Comment (jesperpedersen) [inline src/include/security.h:158]: We don't need this function

File: src/libpgmoneta/security.c [hunk 2]
```diff
@@ -3050,6 +3051,130 @@ pgmoneta_generate_string_hmac_sha256_hash(char* key, int key_length, char* value
...
+      crc32val = crc32_tab[((unsigned char) crc32val) ^
+                           ((unsigned char) buffer[i])] ^
+                 (crc32val >> 8);
+   }
+   *crc_buf = crc32val;
+   return 0;
+}
+
+int
+pgmoneta_create_crc32c_buffer_avx(char* buffer, size_t size, uint32_t* crc_buf)
```
Comment (jesperpedersen) [inline src/libpgmoneta/security.c:3126]: This need to be folded into the core crc32c function

File: src/libpgmoneta/security.c [hunk 3]
```diff
@@ -3050,6 +3051,130 @@ pgmoneta_generate_string_hmac_sha256_hash(char* key, int key_length, char* value
...
+
+   while ((read_bytes = fread(read_buf, 1, sizeof(read_buf), file)) > 0)
+   {
+      pgmoneta_create_crc32c_buffer(read_buf, sizeof(read_buf), &crc_buf);
+   }
+
+   fclose(file);
+
+   crc_string = malloc(9);
+   memset(crc_string, 0, 9);
```
Comment (jesperpedersen) [inline src/libpgmoneta/security.c:3167]: crc_string can be `NULL`

**Review by jesperpedersen — CHANGES_REQUESTED** [review]

**Comment by jesperpedersen** [comment]
You need to check for SSE 4.2 in `CMakeLists.txt` and define a constant that you can use in an `#ifdef`

File: src/libpgmoneta/security.c (hunk 3 above)
Comment (shikharish) [reply src/libpgmoneta/security.c:3167]: Wont that lead to a segmentation fault when I use `sprintf`?

**Comment by jesperpedersen** [comment]
See CI, and remember to squash

**Comment by shikharish** [comment]
@jesperpedersen The implementation is complete. Have verified the hash values with other tools like [rhash](https://github.com/rhash/RHash) and [gsutil](https://github.com/GoogleCloudPlatform/gsutil).

**Comment by shikharish** [comment]
I tested the code on a 810MB text file. Using table lookup it took 2.446s and with SSE it took 0.260s.

File: src/libpgmoneta/security.c (hunk 3 above)
Comment (jesperpedersen) [reply src/libpgmoneta/security.c:3203]:
```
if (crc_string == NULL)
{
   goto error;
}
```

and then

```
error:
   return 1;

File: src/libpgmoneta/security.c [hunk 4]
```diff
@@ -3060,6 +3064,153 @@ pgmoneta_generate_string_hmac_sha256_hash(char* key, int key_length, char* value
...
+   FILE* file = NULL;
+   char read_buf[16384];
+   unsigned long read_bytes = 0;
+   char* crc_string;
+   uint32_t crc_buf = 0;
+
+   file = fopen(path, "rb");
+   if (file == NULL)
+   {
+      return 1;
```
Comment (jesperpedersen) [inline src/libpgmoneta/security.c:3190]:
```
g;oto error;
```

File: src/libpgmoneta/security.c [hunk 5]
```diff
@@ -3060,6 +3064,153 @@ pgmoneta_generate_string_hmac_sha256_hash(char* key, int key_length, char* value
...
+   crc_string = malloc(9);
+   memset(crc_string, 0, 9);
+
+   for (int i = 0; i < 8; i++)
+   {
+      sprintf(crc_string, "%08x", crc_buf);
+   }
+
+   *crc = crc_string;
+   return 0;
```
Comment (jesperpedersen) [inline src/libpgmoneta/security.c:3211]: You are missing `fclose(file)`

File: src/include/security.h [hunk 6]
```diff
@@ -148,6 +148,22 @@ int pgmoneta_generate_string_hmac_sha256_hash(char* key, int key_length, char* v
...
                                               int* hmac_length);
 
+/**
+ * Generate CRC32C for a buffer
+ * @param buffer The buffer.
+ * @param size The size of the buffer.
+ * @param crc_buff The hash value.
+ * @return 0 upon success, otherwise 1.
+ */
+int pgmoneta_create_crc32c_buffer(char* buffer, size_t size, uint32_t* crc_buf);
```
Comment (jesperpedersen) [inline src/include/security.h:158]: `void*` would be better in the signature

File: src/include/security.h [hunk 7]
```diff
@@ -148,6 +148,22 @@ int pgmoneta_generate_string_hmac_sha256_hash(char* key, int key_length, char* v
                                               int value_length, unsigned char** hmac,
                                               int* hmac_length);
 
+/**
+ * Generate CRC32C for a buffer
+ * @param buffer The buffer.
```
Comment (jesperpedersen) [inline src/include/security.h:153]: Remove the `.`

**Review by jesperpedersen — CHANGES_REQUESTED** [review]

File: src/libpgmoneta/security.c (hunk 5 above)
Comment (shikharish) [reply src/libpgmoneta/security.c:3211]: It is present a few lines above.

**Comment by jesperpedersen** [comment]
```
cat /proc/cpuinfo
...
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd cppc arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip rdpid overflow_recov succor smca sev sev_es
```

but

```
-- Performing Test HAVE_SSE42
-- Performing Test HAVE_SSE42 - Failed
```

**Comment by jesperpedersen** [comment]
CI...

**Comment by shikharish** [comment]
@jesperpedersen The test is now only failing for macos. I dont have any experience with macos. Can you please help here?

**Comment by jesperpedersen** [comment]
You can start with `#ifdef HAVE_LINUX` - but test with your FreeBSD installation as well...

**Comment by shikharish** [comment]
@jesperpedersen The macos Actions uses arm64 which does not has SSE4.2 instructions. So the `HAVE_SSE42` should fail. Have updated `CMakeLists.txt` to clear the residual `-msse4.2` flag which was failing the test.

**Comment by jesperpedersen** [comment]
Merged.

Thanks for your contribution !

**Pull request merged**